src/threaddeadlock.cpp

```diff
@@ -0,0 +1,718 @@
+        }
+    }
+    else //  !isExclusive
+    {
+        auto it = lockdata.readlockswaiting.find(c);
+        if (it == lockdata.readlockswaiting.end())
+        {
+            std::set<uint64_t> holders;
+            holders.emplace(tid);
+            lockdata.readlockswaiting.emplace(c, holders);
+        }
+        else
+        {
+            it->second.emplace(tid);
+        }
+    }
+}
+
+void SetWaitingToHeld(void *c, bool isExclusive)
+{
+    std::lock_guard<std::mutex> lock(lockdata.dd_mutex);
+    const uint64_t tid = getTid();
+    if (isExclusive)
+    {
+        auto it = lockdata.writelockswaiting.find(c);
+        if (it == lockdata.writelockswaiting.end())
+        {
+            return;
+        }
+        else
+        {
+            it->second.erase(tid);
+            auto iter = lockdata.writelocksheld.find(c);
+            if (iter == lockdata.writelocksheld.end())
+            {
+                std::set<uint64_t> holders;
+                holders.emplace(tid);
+                lockdata.writelocksheld.emplace(c, holders);
+            }
+            else
+            {
+                iter->second.emplace(tid);
+            }
+        }
+    }
+    else //  !isExclusive
+    {
+        auto it = lockdata.readlockswaiting.find(c);
+        if (it == lockdata.readlockswaiting.end())
+        {
+            return;
+        }
+        else
+        {
+            it->second.erase(tid);
+            auto iter = lockdata.readlocksheld.find(c);
+            if (iter == lockdata.readlocksheld.end())
+            {
+                std::set<uint64_t> holders;
+                holders.emplace(tid);
+                lockdata.readlocksheld.emplace(c, holders);
+            }
+            else
+            {
+                iter->second.emplace(tid);
+            }
+        }
+    }
+
+    auto itheld = lockdata.locksheldbythread.find(tid);
+    if (itheld != lockdata.locksheldbythread.end())
+    {
+        for (auto rit = itheld->second.rbegin(); rit != itheld->second.rend(); ++rit)
+        {
+            if (rit->first == c)
+            {
+                rit->second.ChangeWaitingToHeld();
+                break;
+            }
+        }
+    }
+}
+
+// c = the cs
+// isExclusive = is the current lock exclusive, for a recursive mutex (CCriticalSection) this value should always be
+// true
+void push_lock(void *c, const CLockLocation &locklocation, LockType type, bool isExclusive, bool fTry)
+{
+    std::lock_guard<std::mutex> lock(lockdata.dd_mutex);
+
+    LockStackEntry now = std::make_pair(c, locklocation);
+    // tid of the originating request
+    const uint64_t tid = getTid();
+    // If this is a blocking lock operation, we want to make sure that the locking order between 2 mutexes is consistent
+    // across the program
+    if (fTry)
+    {
+        // a try lock will either get it, or it wont. so just add it.
+        // if we dont get the lock this will be undone in destructor
```

Comment:
which destructor?

Reply:
the destructor of the read or write block for whichever critical section made this function call.

i will edit it to say this